platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/model/impl/stack/DecoratedStackImpl.java

```diff
@@ -0,0 +1,335 @@
+            return true;
+        }
+        if (!(obj instanceof DecoratedStackImpl)) {
+            return false;
+        }
+        DecoratedStackImpl other = (DecoratedStackImpl)obj;
+        return Objects.equals(id, other.id) &&
+               Objects.equals(name, other.name) &&
+               Objects.equals(description, other.description) &&
+               Objects.equals(creator, other.creator) &&
+               Objects.equals(scope, other.getScope()) &&
+               Objects.equals(getTags(), other.getTags()) &&
+               Objects.equals(getWorkspaceConfig(), other.getWorkspaceConfig()) &&
+               Objects.equals(getSource(), other.getSource()) &&
+               Objects.equals(getComponents(), other.getComponents()) &&
+               Objects.equals(getIcon(), other.getIcon()) &&
+               Objects.equals(permissions, other.permissions);
+    }
+
+    @Override
+    public int hashCode() {
+        int hash = 7;
+        hash = 31 * hash + Objects.hashCode(id);
+        hash = 31 * hash + Objects.hashCode(name);
+        hash = 31 * hash + Objects.hashCode(description);
+        hash = 31 * hash + Objects.hashCode(scope);
+        hash = 31 * hash + Objects.hashCode(creator);
+        hash = 31 * hash + Objects.hashCode(getTags());
+        hash = 31 * hash + Objects.hashCode(getWorkspaceConfig());
+        hash = 31 * hash + Objects.hashCode(getSource());
+        hash = 31 * hash + Objects.hashCode(getComponents());
+        hash = 31 * hash + Objects.hashCode(getIcon());
+        hash = 31 * hash + Objects.hashCode(permissions);
+        return hash;
+    }
+
+    @Override
+    public String toString() {
+        return "DecoratedStackImpl{id='" + id +
+               "', name='" + name +
+               "', description='" + description +
+               "', scope='" + scope +
+               "', creator='" + creator +
+               "', tags='" + tags + "'" +
+               ", workspaceConfig=" + workspaceConfig +
+               ", stackSource=" + source +
+               ", components=" + components +
+               ", iconContent=" + icon +
+               ", permission=" + permissions +
+               "}";
+    }
+
+    public static class StackBuilder {
+
+        private String               id;
+        private String               name;
+        private String               description;
+        private String               scope;
+        private String               creator;
+        private List<String>         tags;
+        private WorkspaceConfig      workspaceConfig;
+        private StackSource          source;
+        private List<StackComponent> components;
+        private StackIcon            icon;
+        private Permissions          permissions;
+
+        public StackBuilder generateId() {
+            id = NameGenerator.generate("stack", 16);
+            return this;
+        }
+
+        public StackBuilder setId(String id) {
+            if (id == null) {
```

Comment:
This logic is really implicit and doesn't make any sense for me, because you don't need to call `setId(null)` when you have `generateId` method